app/src/main/java/org/groundplatform/android/ui/map/gms/CachingUpscalingTileProvider.kt

```diff
@@ -0,0 +1,174 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.groundplatform.android.ui.map.gms
+
+import android.graphics.Bitmap
+import android.graphics.BitmapFactory
+import android.graphics.Canvas
+import android.graphics.Paint
+import android.graphics.Rect
+import android.util.LruCache
+import androidx.core.graphics.createBitmap
+import com.google.android.gms.maps.model.Tile
+import com.google.android.gms.maps.model.TileProvider
+import java.io.ByteArrayOutputStream
+import timber.log.Timber
+
+/**
+ * A [TileProvider] that provides upscaled map tiles beyond the available data zoom level.
+ *
+ * This provider attempts to synthesize higher-zoom tiles by:
+ * - Fetching the base tile at [zoomThreshold]
+ * - Cropping the appropriate quadrant
+ * - Upscaling it using bilinear filtering to a 256×256 PNG
+ *
+ * **Performance Note:** The upscaling operation performed by [synthesizeUpscaledTileBytes] is **CPU
+ * and memory intensive** due to bitmap decoding and scaling. Callers should ensure this provider
+ * runs in a background coroutine with an appropriate [CoroutineDispatcher] (e.g. `Dispatchers.IO`)
+ * to prevent blocking the main thread or UI rendering.
+ *
+ * Typical usage:
+ * ```kotlin
+ * val provider = CachingUpscalingTileProvider(source, dataMaxZoom)
+ * val overlay = map.addTileOverlay(
+ *     TileOverlayOptions().tileProvider(provider)
+ * )
+ * ```
+ *
+ * @param source The underlying [TileProvider] that supplies base tiles up to [zoomThreshold].
+ * @param zoomThreshold The maximum zoom level for which base tiles are available. Tiles beyond this
+ *   level will be synthesized by cropping and upscaling.
+ * @param tileSize The size (in pixels) of each output tile. Defaults to [DEFAULT_TILE_SIZE] (256
+ *   px).
+ * @param maxCacheBytes The maximum in-memory cache size in bytes for storing upscaled tiles.
+ */
+class CachingUpscalingTileProvider(
+  private val source: TileProvider,
+  private val zoomThreshold: Int,
+) : TileProvider {
+
+  private val cache =
+    object : LruCache<String, ByteArray>(DEFAULT_CACHE_SIZE_BYTES) {
+      override fun sizeOf(key: String, value: ByteArray) = value.size
+    }
+
+  // Returning early helps prevent deep nesting and keeps the flow simple.
+  override fun getTile(x: Int, y: Int, zoom: Int): Tile {
+    val cacheKey = "$zoom/$x/$y"
+
+    val tile =
+      when {
+        zoom <= zoomThreshold -> {
+          source.getTile(x, y, zoom)
+        }
+
+        cache.get(cacheKey) != null -> {
+          Tile(DEFAULT_TILE_SIZE, DEFAULT_TILE_SIZE, cache.get(cacheKey))
+        }
+
+        else -> {
+          val upscaledBytes = synthesizeUpscaledTileBytes(x, y, zoom)
```

Comment:
Can we first fetch the tile from the source and only call upscale function if at all a tile exists?
Also, let's pass the ByteArray to the function to keep the function single purpose i.e. upscaling and not fetching + upscaling.